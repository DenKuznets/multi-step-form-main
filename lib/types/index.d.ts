export interface Step {
    valid: boolean;
    dirty: boolean;
}

export interface PersonalInfoStep extends Step {
    value: {
        name: string;
        email: string;
        phone: string;
    };
}

export interface SelectPlanStep extends Step {
    value: {
        plan: string;
        subscribtion: string;
    };
}

export interface AddonsStep extends Step {
    value: {
        onlineService: boolean;
        largerStorage: boolean;
        customProfile: boolean;
    };
}

export interface SummaryStep extends Step {}
