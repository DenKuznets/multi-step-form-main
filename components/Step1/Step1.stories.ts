import type { Meta, StoryObj } from '@storybook/react';
import  Step1  from './Step1';

const meta = {
  title: 'Step1',
  component: Step1,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'fullscreen',
  },
} satisfies Meta<typeof Step1>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
