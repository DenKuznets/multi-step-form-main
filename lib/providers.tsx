'use client';

/* Core */
import { Provider } from 'react-redux';
import { setupStore } from './redux/store';
import { PropsWithChildren } from 'react';

export const Providers = ({ children }: PropsWithChildren<{}>) => {
    return (
        <Provider store={setupStore()}>
                {children}
        </Provider>
    );
};
