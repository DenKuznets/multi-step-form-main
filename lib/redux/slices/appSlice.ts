import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from '../store';
import {
    AddonsStep,
    PersonalInfoStep,
    SelectPlanStep,
    Step,
    SummaryStep
} from '@/lib/types';

export enum Steps {
    personalInfo = 1,
    selectPlan,
    addons,
    summary
}

export const PLANS = {
    ARCADE: 'arcade',
    ADVANCED: 'advanced',
    PRO: 'pro'
};

export const SUBSCRIBTION = {
    MONTHLY: 'monthly',
    YEARLY: 'yearly'
};

type stepsType = PersonalInfoStep | SelectPlanStep | AddonsStep | SummaryStep;

// Define a type for the slice state
interface AppState {
    currentStep: number;
    steps: stepsType[]
}

// Define the initial state using that type
const initialState: AppState = {
    currentStep: Steps.personalInfo,
    steps: [
        {
            name:"YOUR INFO",
            valid: false,
            dirty: true,
            value: {
                name: '',
                email: '',
                phone: ''
            }
        },
        {
            name:"SELECT PLAN",
            valid: false,
            dirty: false,
            value: {
                plan: PLANS.ARCADE,
                subscribtion: SUBSCRIBTION.MONTHLY
            }
        },
        {
            name:"ADD-ONS",
            valid: false,
            dirty: true,
            value: {
                cuustomProfile: false,
                largerStorage: false,
                onlineService: false
            }
        },
        {
            name:"SUMMARY",
            valid: false,
            dirty: false
        }
    ]
};

export const appSlice = createSlice({
    name: 'app',
    // `createSlice` will infer the state type from the `initialState` argument
    initialState,
    reducers: {
        increment: (state) => {
            // state.valu
        },
        // Use the PayloadAction type to declare the contents of `action.payload`
        incrementByAmount: (state, action: PayloadAction<number>) => {
            // state.value += action.payload;
        }
    }
});

export const { increment, incrementByAmount } = appSlice.actions;

// Other code such as selectors can use the imported `RootState` type
export const selectCurrentStep = (state: RootState) => state.app.currentStep;
export const selectSteps = (state: RootState) => state.app.steps;

export default appSlice.reducer;
