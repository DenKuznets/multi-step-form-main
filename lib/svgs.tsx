export const Checkmark = ({className }:{className:string}) => (
    <svg
        xmlns="http://www.w3.org/2000/svg"
        width={20}
        height={20}
        fill="none"
        viewBox="0 0 24 24"
        className={className}
    >
        <path
            fill="hsl(243, 100%, 62%)"
            fillRule="evenodd"
            d="M2 4.5A2.5 2.5 0 0 1 4.5 2h15A2.5 2.5 0 0 1 22 4.5v15a2.5 2.5 0 0 1-2.5 2.5h-15A2.5 2.5 0 0 1 2 19.5v-15Zm16.787 5.075a1 1 0 0 0-.003-1.414l-.709-.705a1 1 0 0 0-1.414.003l-5.772 5.796L7.57 9.98a1 1 0 0 0-1.415.01l-.702.712a1 1 0 0 0 .01 1.414L10.2 16.79a1 1 0 0 0 1.411-.007l7.176-7.207Z"
            clipRule="evenodd"
        />
    </svg>
);
