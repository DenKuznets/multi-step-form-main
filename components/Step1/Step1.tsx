'use client';
import { useForm, SubmitHandler } from 'react-hook-form';
import { useAppDispatch, useAppSelector } from '@/lib/redux/hooks';
import { selectName } from '@/lib/redux/slices/appSlice';
import { selectPhone } from '@/lib/redux/slices/appSlice';
import { selectEmail } from '@/lib/redux/slices/appSlice';
import { useRouter } from 'next/navigation';
import { setName, setEmail, setPhone } from '@/lib/redux/slices/appSlice';

interface FormValues {
    name: string;
    phone: string;
    email: string;
}

const Step1 = () => {
    const router = useRouter();
    const name = useAppSelector(selectName);
    const phone = useAppSelector(selectPhone);
    const email = useAppSelector(selectEmail);
    const dispatch = useAppDispatch();
    const { formState, register, handleSubmit } = useForm<FormValues>({
        defaultValues: {
            name: name,
            email: email,
            phone: phone
        }
    });
    const { errors } = formState;
    const onSubmit: SubmitHandler<FormValues> = (data) => {
        console.log('submit', data);
        dispatch(setName(data.name));
        dispatch(setPhone(data.phone));
        dispatch(setEmail(data.email));
        router.push('/select-plan');
    };

    return (
        <form
            noValidate
            className="flex w-full flex-col gap-2 sm:gap-4"
            aria-label="personal info form"
            onSubmit={handleSubmit(onSubmit)}
        >
            <div className="relative">
                <p className="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]">
                    {errors.name?.message}
                </p>
                <label htmlFor="name" className={`label`}>
                    Name
                </label>
                <input
                    id="name"
                    className={`input  ${
                        errors.name ? 'outline-strawberryRed' : ''
                    }`}
                    placeholder="e.g. Stephen King"
                    {...register(`name`, {
                        required: 'This field is required'
                    })}
                />
            </div>
            <div className="relative">
                <p className="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]">
                    {errors.email?.message}
                </p>
                <label htmlFor="email" className="label">
                    Email Address
                </label>
                <input
                    id="email"
                    placeholder="e.g. stephenking@lorem.com"
                    className={`input  ${
                        errors.email ? 'outline-strawberryRed' : ''
                    }`}
                    {...register(`email`, {
                        required: 'This field is required',
                        pattern: {
                            value: /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/g,
                            message: 'Invalid email'
                        }
                    })}
                />
            </div>
            <div className="relative">
                <p className="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]">
                    {errors.phone?.message}
                </p>
                <label htmlFor="phone" className="label">
                    Phone Number
                </label>
                <input
                    id="phone"
                    className={`input  ${
                        errors.phone ? 'outline-strawberryRed' : ''
                    }`}
                    placeholder="e.g. +1 234 567 890"
                    {...register(`phone`, {
                        required: 'This field is required'
                    })}
                />
            </div>

            <div className="fixed bottom-0 left-0 flex w-full justify-end bg-white p-4 sm:absolute sm:mb-4 sm:p-0">
                <button className="btn btn-next ">next step</button>
            </div>
        </form>
    );
};

export default Step1;
