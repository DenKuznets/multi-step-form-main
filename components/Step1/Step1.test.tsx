import '@testing-library/jest-dom';
// import Step1 from '../../app/[stepName]/page';
import {
    renderWithProviders,
    screen,
    within
} from '@/lib/redux/utils-for-tests';
import userEvent from '@testing-library/user-event';
import { initialState } from '@/lib/redux/slices/appSlice';
import { produce } from 'immer';
import { PersonalInfoStep } from '@/lib/types';
import Step1 from './Step1';

test('renders correctly', () => {
    renderWithProviders(
        <Step1 />
    );

    const form = screen.getByRole('form', { name: /personal info form/i });
    const nameLabel = within(form).getByText(/name/i);
    const emailLabel = within(form).getByText(/email address/i);
    const phoneLabel = within(form).getByText(/phone number/i);
    const nameInput = screen.getByRole('textbox', { name: /name/i });
    const emailInput = screen.getByRole('textbox', { name: /email address/i });
    const phoneInput = screen.getByRole('textbox', { name: /phone number/i });
    const nextStepButton = screen.getByRole('button', {
        name: /next step/i
    });

    expect(form).toBeInTheDocument();
    expect(nameLabel).toBeInTheDocument();
    expect(emailLabel).toBeInTheDocument();
    expect(phoneLabel).toBeInTheDocument();
    expect(nameInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(phoneInput).toBeInTheDocument();
    expect(nextStepButton).toBeInTheDocument();
});

test('required message pops up at empty required fields when submit button clicked', async () => {
    const user = userEvent.setup();
    renderWithProviders(<Step1 />);
    const nextStepButton = screen.getByRole('button', {
        name: /next step/i
    });
    await user.click(nextStepButton);
    const requiredMessages = screen.getAllByText('This field is required');
    expect(requiredMessages).toHaveLength(3);
});

test('invalid email message pops up at incorrectly filled email field when submit button clicked', async () => {
    const user = userEvent.setup();
    const newState = produce(initialState, (drafState) => {
        const step = drafState.steps[0] as PersonalInfoStep;
        step.value.email = 'asdas';
    });

    renderWithProviders(<Step1 />, {
        preloadedState: {
            app: {
                ...newState
            }
        }
    });
    const nextStepButton = screen.getByRole('button', {
        name: /next step/i
    });
    await user.click(nextStepButton);
    const requiredMessages = screen.getByText('Invalid email');
    expect(requiredMessages).toBeInTheDocument();
});

test('renders Step1 unchanged', () => {
    const { container } = renderWithProviders(<Step1 />);
    expect(container).toMatchSnapshot();
});