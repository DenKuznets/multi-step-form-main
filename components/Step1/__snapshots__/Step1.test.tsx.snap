// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Step1 unchanged 1`] = `
<div>
  <form
    aria-label="personal info form"
    class="flex w-full flex-col gap-2 sm:gap-4"
    novalidate=""
  >
    <div
      class="relative"
    >
      <p
        class="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]"
      />
      <label
        class="label"
        for="name"
      >
        Name
      </label>
      <input
        class="input  "
        id="name"
        name="name"
        placeholder="e.g. stephenking@lorem.com"
      />
    </div>
    <div
      class="relative"
    >
      <p
        class="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]"
      />
      <label
        class="label"
        for="email"
      >
        Email Address
      </label>
      <input
        class="input  "
        id="email"
        name="email"
        placeholder="e.g. Stephen King"
      />
    </div>
    <div
      class="relative"
    >
      <p
        class="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]"
      />
      <label
        class="label"
        for="phone"
      >
        Phone Number
      </label>
      <input
        class="input  "
        id="phone"
        name="phone"
        placeholder="e.g. +1 234 567 890"
      />
    </div>
    <div
      class="fixed bottom-0 left-0 flex w-full justify-end bg-white p-4 sm:absolute sm:p-0 sm:mb-4"
    >
      <button
        class="btn btn-next "
      >
        next step
      </button>
    </div>
  </form>
</div>
`;
